license_manager/service/license.go

```diff
@@ -0,0 +1,141 @@
+package service
+
+import (
+	"strings"
+	"time"
+
+	lmpb "github.com/enfabrica/enkit/license_manager/proto"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// license manages allocations and queued invocations for a single license
+// type.
+type license struct {
+	totalAvailable int                    // Constant total number of licenses available for invocations.
+	queue          []*invocation          // List of invocations waiting for a license, in FIFO order.
+	allocations    map[string]*invocation // Map of invocation ID to invocation data for an allocated license.
```

Comment:
code looks good, so no required change: but while reviewing the code, I was wondering if it made sense to have a single `map[string]*invocation` containing both allocated and non allocated invocations.

The invocation struct then gains a `prev, next *invocation` pointer, and the `license` struct gains a `first, last *invocation`. If the invocation is not allocated, then prev and next link it into a linked list starting from first. If the invocation is allocated, then either those two pointers are nil, or there's some other bit indicating it is allocated.

Reply:
Sounds like that can work as well - I'd have to think more about which is better. This is more straightforward to me (obviously because it's the first thing my mind jumped to) mainly because the data structures are more concrete (instead of storing them all in a map and then constructing a virtual organization using pointers); I don't think our numbers are high enough to worry about e.g. queue resizing over simplicity of implementation.

Reply:
SGTM